### PR TITLE
Refactor persistence

### PR DESCRIPTION
**Pull request by samgreen**
This migrates all file handling and persistence functionality in to an `MPPersistence` helper. This alos adds a new test suite focused exclusively on persistence.

Tests Status:
screen shot 2016-06-13 at 4 26 02 pm

Code Coverage Report:
screen shot 2016-06-13 at 4 26 26 pm

### CODE REVIEW

File: Mixpanel/Mixpanel.m [hunk 1]
```diff
@@ -389,10 +392,8 @@ - (void)identify:(NSString *)distinctId
                 r[@"$distinct_id"] = distinctId;
                 [self.peopleQueue addObject:r];
             }
+            [self.persistence archivePeopleQueue:self.peopleQueue];
             [self.people.unidentifiedQueue removeAllObjects];
-            [self archivePeople];
-        }
-        if ([Mixpanel inBackground]) {
```
Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: why remove this check?

File: Mixpanel/Mixpanel.m [hunk 2]
```diff
@@ -469,11 +470,8 @@ - (void)track:(NSString *)event properties:(NSDictionary *)properties
             [self.eventsQueue removeObjectAtIndex:0];
         }
         
-        if ([Mixpanel inBackground]) {
```
Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: why did we do this before?

Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: I mean, why did we do `[self archiveEvents]` twice here?

File: Mixpanel/Mixpanel.m [hunk 3]
```diff
@@ -856,80 +801,23 @@ - (void)archiveProperties
...
 
 - (void)unarchive
 {
-    [self unarchiveEvents];
-    [self unarchivePeople];
+    self.eventsQueue = [self.persistence unarchiveEventQueue];
+    self.peopleQueue = [self.persistence unarchivePeopleQueue];
+    
     [self unarchiveProperties];
-    [self unarchiveVariants];
```
Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: what about `unarchiveVariants` and `unarchiveEventBindings`? we don't need them any more?

Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: Oh..so they are done in the `unarchiveProperties`. but why this change? since we have separate logic to handle things like `unarchiveVariants`, do you think it makes more sense to separate them here as well?

File: Mixpanel/Mixpanel.m [hunk 4]
```diff
@@ -2072,22 +1951,24 @@ - (void)addPeopleRecordToQueueWithAction:(NSString *)action andProperties:(NSDic
...
                 MixpanelDebug(@"%@ queueing people record: %@", self.mixpanel, r);
-                [strongMixpanel.peopleQueue addObject:r];
-                if (strongMixpanel.peopleQueue.count > 500) {
-                    [strongMixpanel.peopleQueue removeObjectAtIndex:0];
+                [peopleQueue addObject:r];
+                if (peopleQueue.count > 500) {
+                    [peopleQueue removeObjectAtIndex:0];
                 }
+                MPPersistence *persistence = strongMixpanel.persistence;
+                [persistence archivePeopleQueue:peopleQueue];
```
Comment (yinfeiru) [inline Mixpanel/Mixpanel.m]: why do we move this logic inside this `if`? I think we want to archive people if `people.identify` isn't called yet.

File: Mixpanel/Mixpanel.m (hunk 2 above)
Comment (samgreen) [inline Mixpanel/Mixpanel.m]: This was a bug, it meant we would archive twice if we were in the background.

File: Mixpanel/Mixpanel.m (hunk 3 above)
Comment (samgreen) [inline Mixpanel/Mixpanel.m]: Trying to keep the refactoring to a minimum. There are many functions that call `unarchiveProperties`, I don't want to go change all of those call sites to also call `unarchiveVariants` and `unarchiveEventBindings`

File: Mixpanel/Mixpanel.m (hunk 4 above)
Comment (samgreen) [inline Mixpanel/Mixpanel.m]: We're archiving the people queue which is only valid if they have called `identify:`, if they haven't yet identified the user then we need to saved the `unidentifiedQueue` which contains events that still need their `distinctID` updated.
